controllers/observability_reconciler.go

```diff
@@ -0,0 +1,111 @@
+package controllers
+
+import (
+	"context"
+	"sync"
+
+	kuadrantv1beta1 "github.com/kuadrant/kuadrant-operator/api/v1beta1"
+	"github.com/kuadrant/kuadrant-operator/pkg/utils"
+	"github.com/kuadrant/policy-machinery/controller"
+	"github.com/kuadrant/policy-machinery/machinery"
+	monitoringv1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	monclient "github.com/prometheus-operator/prometheus-operator/pkg/client/versioned"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/utils/ptr"
+)
+
+type ObservabilityReconciler struct {
+	Client     *dynamic.DynamicClient
+	restMapper meta.RESTMapper
+	MonClient  monclient.Interface
+}
+
+func NewObservabilityReconciler(client *dynamic.DynamicClient, rm meta.RESTMapper, mc monclient.Interface) *ObservabilityReconciler {
+	return &ObservabilityReconciler{
+		Client:     client,
+		restMapper: rm,
+		MonClient:  mc,
+	}
+}
+
+func (r *ObservabilityReconciler) Subscription() *controller.Subscription {
+	return &controller.Subscription{ReconcileFunc: r.Reconcile, Events: []controller.ResourceEventMatcher{
+		{Kind: ptr.To(kuadrantv1beta1.KuadrantGroupKind)},
+	},
+	}
+}
+
+func (r *ObservabilityReconciler) Reconcile(ctx context.Context, _ []controller.ResourceEvent, topology *machinery.Topology, _ error, _ *sync.Map) error {
+	logger := controller.LoggerFromContext(ctx).WithName("ObservabilityReconciler")
+	logger.Info("reconciling observability", "status", "started")
+	defer logger.Info("reconciling observability", "status", "completed")
+
+	kObj := GetKuadrantFromTopology(topology)
+	if kObj == nil {
+		return nil
+	}
+
+	// Check if observability is enabled
+	if !kObj.Spec.Observability.Enable {
+		logger.Info("observability enable is false")
+		// TODO: Remove monitors if they exist
+		return nil
+	}
+	logger.Info("observability enable is true")
+
+	// Check if monitoring CRDs exist
+	if ok, err := utils.IsCRDInstalled(r.restMapper, monitoringv1.SchemeGroupVersion.Group, monitoringv1.ServiceMonitorsKind, monitoringv1.SchemeGroupVersion.Version); !ok || err != nil {
+		logger.Info("ServiceMonitor CRD not installed")
+		return nil
+	}
+	logger.Info("ServiceMonitor CRD is installed")
```

Comment:
nitpick: I would make this debug level, `logger.V(1).Info("ServiceMonitor CRD is installed")`